adam_core/backend/backend.py

```diff
@@ -0,0 +1,168 @@
+import copy
+import logging
+import multiprocessing as mp
+from abc import ABC, abstractmethod
+from typing import Union
+
+from astropy.time import Time
+
+from ..orbits import Orbits
+from ..utils.indexable import concatenate
+from ..utils.multiprocessing import _check_parallel
+
+logger = logging.getLogger(__name__)
+
+
+def propagation_worker(orbits: Orbits, times: Time, backend: "Backend") -> Orbits:
+    propagated = backend._propagate_orbits(orbits, times)
+    return propagated
+
+
+def ephemeris_worker(orbits: Orbits, observers, backend: "Backend"):
+    ephemeris = backend._generate_ephemeris(orbits, observers)
+    return ephemeris
+
+
+class Backend(ABC):
+    @abstractmethod
+    def _propagate_orbits(self, orbits: Orbits, times: Time) -> Orbits:
+        """
+        Propagate orbits to times.
+
+        THIS FUNCTION SHOULD BE DEFINED BY THE USER.
+        """
+        pass
+
+    def propagate_orbits(
+        self,
+        orbits: Orbits,
+        times: Time,
+        chunk_size: int = 100,
+        num_jobs: int = 1,
+    ) -> Orbits:
+        """
+        Propagate each orbit in orbits to each time in times.
+
+        Parameters
+        ----------
+        orbits : `~adam_core.orbits.orbits.Orbits` (N)
+            Orbits to propagate.
+        times : `~astropy.time.core.Time` (M)
+            Times to which to propagate orbits.
+        chunk_size : int, optional
+            Number of orbits to send to each job.
+        num_jobs : int or "auto", optional
+            Number of jobs to launch. If "auto" then the number of
+            jobs will be equal to the number of cores on the machine.
+
+        Returns
+        -------
+        propagated : `~adam_core.orbits.orbits.Orbits`
+            Propagated orbits.
+        """
+        parallel, num_workers = _check_parallel(num_jobs)
+        if parallel:
+            orbits_split = list(orbits.yield_chunks(chunk_size))
+            times_duplicated = [copy.copy(times) for i in range(len(orbits_split))]
+            backend_duplicated = [copy.copy(self) for i in range(len(orbits_split))]
+
+            p = mp.Pool(
+                processes=num_workers,
+            )
+
+            propagated_list = p.starmap(
+                propagation_worker,
+                zip(
+                    orbits_split,
+                    times_duplicated,
+                    backend_duplicated,
```

Comment:
Why does the backend need to be copied?

Also - probably worth loudly documenting that backends need to be picklable. `multiprocessing` uses `pickle` to send these things to the other processes. That means that backends can't, for example, have any open file handles or network connections.

Reply:
I removed the copies and added calls to `itertools.repeat` rather than list comprehensions. Hopefully that is cleaner.

Also added a statement that subclasses must be pickleable.